cpp/src/arrow/compute/row/row_encoder_internal.h

```diff
@@ -269,6 +272,190 @@ struct ARROW_EXPORT NullKeyEncoder : KeyEncoder {
   }
 };
 
+template <typename ListType>
+struct ARROW_EXPORT ListKeyEncoder : KeyEncoder {
```

Comment:
Wonder should I put this into `.cc` since it requires a lot

Reply:
Yes, please do. It would be nice to hide most contents from this file into the corresponding `.cc`